File: packages/storage/src/signed-urls.ts
import type { DirectUploadCapability } from "./types.js";

// HMAC-SHA256 signed URL tokens for pre-authorized downloads
// No external dependencies — uses Web Crypto API

export type DownloadUrlPayload = {
  key: string;
  expiresAt: number; // unix timestamp ms
};

function base64urlEncode(data: Uint8Array): string {
  let binary = "";
  for (let i = 0; i < data.length; i++) {
    binary += String.fromCharCode(data[i] as number);
  }
  return btoa(binary).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

function base64urlDecode(str: string): Uint8Array {
  const padded = str.replace(/-/g, "+").replace(/_/g, "/");
  const pad = padded.length % 4;
  const paddedStr = pad === 0 ? padded : padded + "=".repeat(4 - pad);
  const binary = atob(paddedStr);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes;
}

async function importHmacKey(secret: string): Promise<CryptoKey> {
  if (secret.trim() === "") {
    throw new Error("Signing secret must not be blank");
  }
  const keyData = new TextEncoder().encode(secret);
  return crypto.subtle.importKey("raw", keyData, { name: "HMAC", hash: "SHA-256" }, false, [
    "sign",
    "verify",
  ]);
}

async function hmacSign(data: string, secret: string): Promise<Uint8Array> {
  const key = await importHmacKey(secret);
  const sig = await crypto.subtle.sign("HMAC", key, new TextEncoder().encode(data));
  return new Uint8Array(sig);
}

async function hmacVerify(data: string, sig: Uint8Array, secret: string): Promise<boolean> {
  const key = await importHmacKey(secret);
  const signature = new Uint8Array(sig.byteLength);
  signature.set(sig);
  return crypto.subtle.verify("HMAC", key, signature.buffer, new TextEncoder().encode(data));
}

// Serialize payload as JSON, HMAC-SHA256 sign with secret
// Return base64url-encoded `${base64url(json)}.${base64url(sig)}`
export async function signDownloadUrl(
  payload: DownloadUrlPayload,
  secret: string,
): Promise<string> {
  const json = JSON.stringify(payload);
  const encodedPayload = base64urlEncode(new TextEncoder().encode(json));
  const sig = await hmacSign(encodedPayload, secret);
  const encodedSig = base64urlEncode(sig);
  return `${encodedPayload}.${encodedSig}`;
}

// Parse token, verify signature, check expiry
// Return null if invalid or expired
export async function verifyDownloadUrl(
  token: string,
  secret: string,
): Promise<DownloadUrlPayload | null> {
  if (secret.trim() === "") return null;
  const parts = token.split(".");
  if (parts.length !== 2) return null;

  const [encodedPayload, encodedSig] = parts as [string, string];

  let sigBytes: Uint8Array;
  let payloadBytes: Uint8Array;
  try {
    sigBytes = base64urlDecode(encodedSig);
    payloadBytes = base64urlDecode(encodedPayload);
  } catch {
    return null;
  }

  const valid = await hmacVerify(encodedPayload, sigBytes, secret);
  if (!valid) return null;

  let payload: unknown;
  try {
    payload = JSON.parse(new TextDecoder().decode(payloadBytes));
  } catch {
    return null;
  }

  if (
    typeof payload !== "object" ||
    payload === null ||
    typeof (payload as Record<string, unknown>).key !== "string" ||
    !Number.isSafeInteger((payload as Record<string, unknown>).expiresAt)
  ) {
    return null;
  }

  const typed = payload as DownloadUrlPayload;
  if (Date.now() > typed.expiresAt) return null;

  return typed;
}

export async function generateCapabilityToken(
  capability: DirectUploadCapability,
  secret: string,
): Promise<string> {
  const json = JSON.stringify(capability);
  const encodedPayload = base64urlEncode(new TextEncoder().encode(json));
  const sig = await hmacSign(encodedPayload, secret);
  const encodedSig = base64urlEncode(sig);
  return `${encodedPayload}.${encodedSig}`;
}

export async function verifyCapabilityToken(
  token: string,
  secret: string,
): Promise<DirectUploadCapability | null> {
  if (secret.trim() === "") return null;
  const parts = token.split(".");
  if (parts.length !== 2) return null;

  const [encodedPayload, encodedSig] = parts as [string, string];

  let sigBytes: Uint8Array;
  let payloadBytes: Uint8Array;
  try {
    sigBytes = base64urlDecode(encodedSig);
    payloadBytes = base64urlDecode(encodedPayload);
  } catch {
    return null;
  }

  const valid = await hmacVerify(encodedPayload, sigBytes, secret);
  if (!valid) return null;

  let capability: unknown;
  try {
    capability = JSON.parse(new TextDecoder().decode(payloadBytes));
  } catch {
    return null;
  }

  if (
    typeof capability !== "object" ||
    capability === null ||
    typeof (capability as Record<string, unknown>).key !== "string" ||
    typeof (capability as Record<string, unknown>).contentType !== "string" ||
    !Number.isSafeInteger((capability as Record<string, unknown>).expiresAt)
  ) {
    return null;
  }
  const maxSizeBytes = (capability as Record<string, unknown>).maxSizeBytes;
  if (maxSizeBytes !== undefined && !Number.isSafeInteger(maxSizeBytes)) {
    return null;
  }

  const typed = capability as DirectUploadCapability;
  if (Date.now() > typed.expiresAt) return null;

  return typed;
}

File: py/ventora_storage/src/ventora_storage/signed_urls.py
from __future__ import annotations

import base64
import hashlib
import hmac
import json
import time
from dataclasses import dataclass
from typing import Any, TypeGuard

BASE64URL_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_"


@dataclass(frozen=True)
class DownloadUrlPayload:
    key: str
    expires_at: int


@dataclass(frozen=True)
class DirectUploadInput:
    key: str
    content_type: str
    max_size_bytes: int | None = None
    expires_in: int | None = None


@dataclass(frozen=True)
class DirectUploadCapability:
    key: str
    content_type: str
    expires_at: int
    max_size_bytes: int | None = None


def _base64url_encode(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).decode("ascii").rstrip("=")


def _base64url_decode(value: str) -> bytes:
    if not value or any(char not in BASE64URL_ALPHABET for char in value):
        raise ValueError("invalid base64url")
    padding = "=" * (-len(value) % 4)
    return base64.urlsafe_b64decode(f"{value}{padding}")


def _hmac_signature(encoded_payload: str, secret: str) -> bytes:
    _validate_secret(secret)
    return hmac.new(
        secret.encode("utf-8"),
        encoded_payload.encode("utf-8"),
        hashlib.sha256,
    ).digest()


def _validate_secret(secret: str) -> None:
    if not secret.strip():
        raise ValueError("secret must not be blank")


def _sign_payload(payload: dict[str, object], secret: str) -> str:
    _validate_secret(secret)
    json_payload = json.dumps(payload, separators=(",", ":"))
    encoded_payload = _base64url_encode(json_payload.encode("utf-8"))
    encoded_signature = _base64url_encode(_hmac_signature(encoded_payload, secret))
    return f"{encoded_payload}.{encoded_signature}"


def _verify_payload(token: str, secret: str) -> dict[str, Any] | None:
    try:
        _validate_secret(secret)
    except ValueError:
        return None

    parts = token.split(".")
    if len(parts) != 2:
        return None

    encoded_payload, encoded_signature = parts
    try:
        signature = _base64url_decode(encoded_signature)
        payload_bytes = _base64url_decode(encoded_payload)
    except ValueError:
        return None

    expected_signature = _hmac_signature(encoded_payload, secret)
    if not hmac.compare_digest(signature, expected_signature):
        return None

    try:
        payload = json.loads(payload_bytes.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError):
        return None

    if not isinstance(payload, dict):
        return None
    return payload


def _is_expiration_millis(value: object) -> TypeGuard[int]:
    return isinstance(value, int) and not isinstance(value, bool)


def _is_not_expired(expires_at: int) -> bool:
    return int(time.time() * 1000) <= expires_at


def sign_download_url(payload: DownloadUrlPayload, secret: str) -> str:
    return _sign_payload({"key": payload.key, "expiresAt": payload.expires_at}, secret)


def verify_download_url(token: str, secret: str) -> DownloadUrlPayload | None:
    payload = _verify_payload(token, secret)
    if payload is None:
        return None

    key = payload.get("key")
    expires_at = payload.get("expiresAt")
    if not isinstance(key, str) or not _is_expiration_millis(expires_at):
        return None
    if not _is_not_expired(expires_at):
        return None

    return DownloadUrlPayload(key=key, expires_at=int(expires_at))


def generate_capability_token(capability: DirectUploadCapability, secret: str) -> str:
    payload: dict[str, object] = {
        "key": capability.key,
        "contentType": capability.content_type,
    }
    if capability.max_size_bytes is not None:
        payload["maxSizeBytes"] = capability.max_size_bytes
    payload["expiresAt"] = capability.expires_at
    return _sign_payload(payload, secret)


def verify_capability_token(token: str, secret: str) -> DirectUploadCapability | None:
    payload = _verify_payload(token, secret)
    if payload is None:
        return None

    key = payload.get("key")
    content_type = payload.get("contentType")
    expires_at = payload.get("expiresAt")
    max_size_bytes = payload.get("maxSizeBytes")
    if (
        not isinstance(key, str)
        or not isinstance(content_type, str)
        or not _is_expiration_millis(expires_at)
    ):
        return None
    if max_size_bytes is not None and (
        not isinstance(max_size_bytes, int) or isinstance(max_size_bytes, bool)
    ):
        return None
    if not _is_not_expired(expires_at):
        return None

    return DirectUploadCapability(
        key=key,
        content_type=content_type,
        max_size_bytes=max_size_bytes,
        expires_at=int(expires_at),
    )
